v0.4 changelog (#415)
PennyLaneAI/catalyst: doc/changelog.md

  ```python
  import pennylane as qml

  dev = qml.device("lightning.qubit", wires=2)

  @qml.qjit
  @qml.qnode(dev)
  def circuit(theta):
      qml.Hadamard(wires=0)
      qml.RX(theta, wires=1)
      qml.CNOT(wires=[0, 1])
      return qml.expval(qml.PauliZ(wires=1))
  ```

  ```pycon
  >>> circuit(0.5)  # the first call, compilation occurs here
  array(0.)
  >>> circuit(0.5)  # the precompiled quantum function is called
  array(0.)
  ```

  Currently, PennyLane supports the [Catalyst hybrid compiler](https://github.com/pennylaneai/catalyst)
  with the `qml.qjit` decorator, which directly aliases Catalyst's `catalyst.qjit`.

  In addition to the above `qml.qjit` integration, the following native PennyLane functions can now
  be used with the `qjit` decorator: `qml.adjoint`, `qml.ctrl`, `qml.grad`, `qml.jacobian`,
  `qml.vjp`, `qml.jvp`, and `qml.adjoint`, `qml.while_loop`, `qml.for_loop`, `qml.cond`. These will
  alias to the corresponding Catalyst functions when used within a `qjit` context.

  For more details on these functions, please refer to the
 [PennyLane compiler documentation](https://docs.pennylane.ai/en/stable/introduction/compiling_workflows.html) and

>  [`qml.compiler`](https://docs.pennylane.ai/en/stable/code/qml_compiler.html).

Author: ```suggestion
 [compiler module documentation](https://docs.pennylane.ai/en/stable/code/qml_compiler.html).
```

  the previously-compiled binary, resulting in significant performance improvements.

  ```python
  import pennylane as qml

  dev = qml.device("lightning.qubit", wires=2)

  @qml.qjit
  @qml.qnode(dev)
  def circuit(theta):
      qml.Hadamard(wires=0)
      qml.RX(theta, wires=1)
      qml.CNOT(wires=[0, 1])
      return qml.expval(qml.PauliZ(wires=1))
  ```

  ```pycon
  >>> circuit(0.5)  # the first call, compilation occurs here
  array(0.)
  >>> circuit(0.5)  # the precompiled quantum function is called
  array(0.)
  ```

  Currently, PennyLane supports the [Catalyst hybrid compiler](https://github.com/pennylaneai/catalyst)
  with the `qml.qjit` decorator, which directly aliases Catalyst's `catalyst.qjit`.

  In addition to the above `qml.qjit` integration, the following native PennyLane functions can now
  be used with the `qjit` decorator: `qml.adjoint`, `qml.ctrl`, `qml.grad`, `qml.jacobian`,
  `qml.vjp`, `qml.jvp`, and `qml.adjoint`, `qml.while_loop`, `qml.for_loop`, `qml.cond`. These will
  alias to the corresponding Catalyst functions when used within a `qjit` context.

  For more details on these functions, please refer to the

>   [PennyLane compiler documentation](https://docs.pennylane.ai/en/latest/introduction/compiling_workflows.html).

Author: good catch! https://github.com/PennyLaneAI/pennylane/pull/5009

* Just-in-time compiled functions now support asynchronuous execution of QNodes.
  [(#374)](https://github.com/PennyLaneAI/catalyst/pull/374)
  [(#381)](https://github.com/PennyLaneAI/catalyst/pull/381)

  Simply specify `async_qnodes=True` when using the `@qjit` decorator to enable the async
  execution of QNodes. Currently, asynchronous execution is only supported by

>   `lightning.qubit` and `lightning.kokkos`.

Author: Given https://github.com/PennyLaneAI/catalyst/pull/433 is now merged, I'll resolve this comment - let me know if you want to make any mentions of potential oversubscription issues

      qml.RY(params[1], wires=1)
      qml.CNOT(wires=[0, 1])
      return qml.expval(qml.PauliZ(wires=0))

  @qjit(async_qnodes=True)
  def multiple_qnodes(params):
      x = jnp.sin(params)
      y = jnp.cos(params)
      z = jnp.array([circuit(x), circuit(y)]) # will be executed in parallel
      return circuit(z)
  ```
  ``` pycon
  >>> func(jnp.array([1.0, 2.0]))
  1.0
  ```

  Here, the first two circuit executions will occur in parallel across multiple threads,
  as their execution can occur indepdently.

* Preliminary support for PennyLane transforms has been added.
  [(#280)](https://github.com/PennyLaneAI/catalyst/pull/280)

  ```python
  @qjit
  @qml.transforms.split_non_commuting
  @qml.qnode(dev)
  def circuit(x):
      qml.RX(x,wires=0)
      return [qml.expval(qml.PauliY(0)), qml.expval(qml.PauliZ(0))]
  ```

  ```pycon
  >>> circuit(0.4)
  [array(-0.51413599), array(0.85770868)]
  ```

  Currently, most PennyLane transforms will work with Catalyst
  as long as:
  
  - The circuit does not include any Catalyst-specific features, such
    as Catalyst control flow or measurement,
  
  - The QNode returns only lists of measurement processes,
  

>   - AutoGraph is disabled, and

Author: I think this restriction is related to that thread on Slack. E.g., in this small MVP,

```python
@qml.transform
def my_quantum_transform(tape):
    
    new_operations = []

    for op in tape:
      if op.name == "RX":
          if op.parameters[0] < 0:
              new_operations.append(qml.RY(-op.parameters[0], wires=op.wires))
      else:
          new_operations.append(op)

    new_tape = type(tape)(new_operations, tape.measurements, shots=tape.shots)

    def post_processing_fn(results):
        return results[0]

    return [new_tape], post_processing_fn

@qml.qjit(autograph=True)
@my_quantum_transform
@qml.qnode(dev)
def circuit(x):
    qml.RY(x, wires=0)
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))
```

`AutoGraph=True` will result in this circuit executing without error, but the transform will not be properly applied

>   - AutoGraph is disabled, and

Author: There is no issue, but I can make one!

>   - AutoGraph is disabled, and

Author: Now that we know that `autograph=True` simply ignores all transforms, we should leave this in.

  the previously-compiled binary, resulting in significant performance improvements.

  ```python
  import pennylane as qml

  dev = qml.device("lightning.qubit", wires=2)

  @qml.qjit
  @qml.qnode(dev)
  def circuit(theta):
      qml.Hadamard(wires=0)
      qml.RX(theta, wires=1)
      qml.CNOT(wires=[0, 1])
      return qml.expval(qml.PauliZ(wires=1))
  ```

  ```pycon
  >>> circuit(0.5)  # the first call, compilation occurs here
  array(0.)
  >>> circuit(0.5)  # the precompiled quantum function is called
  array(0.)
  ```

  Currently, PennyLane supports the [Catalyst hybrid compiler](https://github.com/pennylaneai/catalyst)
  with the `qml.qjit` decorator, which directly aliases Catalyst's `catalyst.qjit`.

  In addition to the above `qml.qjit` integration, the following native PennyLane functions can now
  be used with the `qjit` decorator: `qml.adjoint`, `qml.ctrl`, `qml.grad`, `qml.jacobian`,
  `qml.vjp`, `qml.jvp`, and `qml.adjoint`, `qml.while_loop`, `qml.for_loop`, `qml.cond`. These will
  alias to the corresponding Catalyst functions when used within a `qjit` context.

  For more details on these functions, please refer to the

>   [PennyLane compiler documentation](https://docs.pennylane.ai/en/latest/introduction/compiling_workflows.html).

Reviewer: @josh146 The compiler module doc page is also a good target to be linked here. 
Perhaps, we can update this line as follows:
 ```suggestion
  [PennyLane compiler documentation](https://docs.pennylane.ai/en/stable/introduction/compiling_workflows.html) and
  [`qml.compiler`](https://docs.pennylane.ai/en/stable/code/qml_compiler.html).
```

Would you also like to use "stable" instead of "latest" for the docs?

   [1. 1. 1.]]
  ```

  When passing tensor arguments to compiled functions as arguments, the
  `abstracted_axes` keyword argument to the `@qjit` can be used to specify
  which axes of the input arguments should be treated as abstract (and thus
  avoid recompilation).

  For example, without specifying `abstracted_axes`, the following `sum` function
  would recompile each time an array of different size is passed
  as an argument:

  ```pycon
  >>> @qjit
  >>> def sum(arr):
  >>>     return jnp.sum(arr)
  >>> sum(jnp.array([1]))     # Compilation happens here.
  >>> sum(jnp.array([1, 1]))  # And here!
  ```

  By passing `abstracted_axes`, we can specify that the first axes
  of the first argument is to be treated as dynamic during initial compilation:

  ```pycon

>   >>> @qjit(abstracted_axes={0: "n"})

Author: Yep, admittedly I can't think of anything better at the moment! So I'm happy to keep iterating on the UI

  >>> def sum_abstracted(arr):
  >>>     return jnp.sum(arr)
  >>> sum(jnp.array([1]))     # Compilation happens here.
  >>> sum(jnp.array([1, 1]))  # No need to recompile.
  ```

  Note that support for dynamic arrays in control-flow primitives (such as loops),

>   as well as the ability to index dynamic arrays, is not yet supported.

Reviewer 2: I'll close this point since it would take a bit of time to rebase the indexing PR.

* A mitigation dialect (MLIR) was added. It initially contains a Zero Noise Extrapolation (ZNE) operation,
  with a lowering to a global folded circuit.

>   [(#324)](https://github.com/PennyLaneAI/catalyst/pull/324)

Author: TODO once #414 is merged

Author: excellent!

  ```pycon
  >>> dev = qml.device("lightning.qubit", wires=2, shots=100)
  >>> @qjit
  >>> @qml.qnode(dev)
  >>> def circuit(x):
  >>>     qml.RX(x, wires=0)
  >>>     return qml.probs(wires=0)
  >>> circuit(0.54)
  array([0.94, 0.06])
  >>> circuit(0.54)
  array([0.93, 0.07])
  ```

* Catalyst gradient functions `grad`, `jacobian`, `jvp`, and `vjp` can now be invoked from
  outside a `@qjit` context.
  [(#375)](https://github.com/PennyLaneAI/catalyst/pull/375)

  This simplifies the process of writing functions where compilation
  can be turned on and off easily by adding or removing the decorator. The functions dispatch to
  their JAX equivalents when the compilation is turned off.

  ```python
  dev = qml.device("lightning.qubit", wires=2)

  @qml.qnode(dev)
  def circuit(x):
      qml.RX(x, wires=0)
      return qml.expval(qml.PauliZ(0))
  ```

  ```pycon
  >>> grad(circuit)(0.54)  # dispatches to jax.grad
  Array(-0.51413599, dtype=float64, weak_type=True)
  >>> qjit(grad(circuit))(0.54). # differentiates using Catalyst
  array(-0.51413599)
  ```

* All `lightning.qubit` configuration options are now supported via the `qml.device` loader,

>   including Markov Chain Monte Carlo sampling support.

Author: @erick-xanadu is this correct? E.g., is the support for parallel observable measurements via the lightning `batch_obs=True` working?

Reviewer 3: I am not sure `batch_obs=True`. What this line says to me is that `mcmc=True` and `num_burnin` and `kernel_name` options are available and the measurements happen via the `generate_samples_metropolis` function. Maybe @maliasadi can help me understand better if what we are both saying is the same thing?

Reviewer: @josh146  @erick-xanadu 
- `mcmc` : it's Markov Chain Monte Carlo sampling support (it's implemented in the C++ API of Lightning)
- `batch_obs` : it calculates jacobians for chucks of observables (up to my understanding this is mainly in Python and added to the Python API of Lightning)

Anyways, we don't support `batch_obs` at the moment.

Author: ```suggestion
* New `lightning.qubit` configuration options are now supported via the `qml.device` loader,
  including Markov Chain Monte Carlo sampling support.
```

    which updates the status of the device instance from `Active` to `Inactive`. The runtime will reuse
    this deactivated instance instead of creating a new one automatically at runtime in a
    multi-QNode workflow when another device with identical specifications is requested.
    [(#381)](https://github.com/PennyLaneAI/catalyst/pull/381)

  - The `DeviceOp` definition in the Quantum MLIR dialect has been updated to lower a tuple
    of device information `('lib', 'name', 'kwargs')` to a single device initialization call
    `__quantum__rt__device_init(int8_t *, int8_t *, int8_t *)`. This allows the runtime to initialize
    device instances without keeping partial information of the device
    [(#396)](https://github.com/PennyLaneAI/catalyst/pull/396)

* The quantum adjoint compiler routine has been extended to support function calls that affect the
  quantum state within an adjoint region. Note that the function may only provide a single result
  consisting of the quantum register. By itself this provides no user-facing changes, but compiler
  pass developers may now generate quantum adjoint operations around a block of code containing

>   function calls as well as quantum operations and control flow operations.

Author: @rmoyard could you explain this change to me as well?

Reviewer 4: You can use have `func.call` inside `quantum.adjoint` operation and it will work.

```
func.func circuit(...)

quantum.adjoint(qreg) {
    func.call(circuit)
}
```

  [(#353)](https://github.com/PennyLaneAI/catalyst/pull/353)

> * `AllocOp`, `DeallocOp` have now (only) value semantics. In the frontend, the last
>   quantum register is deallocated instead of the first one. This allows to return the quantum
>   register in functions and can be given to another function (useful for quantum transformation).

Reviewer 2: ```suggestion
* The allocation and deallocation operations in MLIR (`AllocOp`, `DeallocOp`) now follow simple
  value semantics for qubit register values, instead of modelling memory in the MLIR trait system.
  Similarly, the frontend generates proper value semantics by deallocating the final register value.
  
  The change enables functions at the MLIR level to accept and return quantum register values,
  which would otherwise not be correctly identified as aliases of existing register values by the
  bufferization system. 
```

Author: thanks :)

>   [(#360)](https://github.com/PennyLaneAI/catalyst/pull/360)

Author: @rmoyard I'm not sure what this changelog entry means 🤔

Reviewer 4: It is a very internal change, before we would always dealloc the first quantum register and that would eliminate some operations that were necessary for the ZNE prs. Now we always dellaoc the last register.

<h3>Breaking changes</h3>

> * Third party devices must now specify a configuration TOML file, in order to specify their

Reviewer 2: ```suggestion
* Third party devices must now provide a configuration TOML file, in order to specify their
```

  supported operations, measurements, and features for Catalyst compatibility. For more information
  please visit the [Custom Devices](https://docs.pennylane.ai/projects/catalyst/en/latest/dev/custom_devices.html) section in our documentation.
  [(#369)](https://github.com/PennyLaneAI/catalyst/pull/369)

<h3>Bug fixes</h3>

* Resolves a bug in the compiler's differentiation engine that results in a segmentation fault
  when [attempting to differentiate non-differentiable quantum operations](https://github.com/PennyLaneAI/catalyst/issues/384).

>   The fix ensures that all current quantum operation types are removed during gradient passes that

Reviewer 2: ```suggestion
  The fix ensures that all existing quantum operation types are removed during gradient passes that
```

>   extract classical from a QNode function. It also adds a verification step that will raise an error

Reviewer 2: ```suggestion
  extract classical code from a QNode function. It also adds a verification step that will raise an error
```

  if a gradient pass cannot successfully eliminate all quantum operations for such functions.
  [(#397)](https://github.com/PennyLaneAI/catalyst/issues/397)

> * Resolves a bug where printing multiple strings from multiple functions with `debug.print` could
>   result in a segfault, the printing garbage values, or accidental printing of adjacent strings.

Reviewer 2: ```suggestion
* Resolves a bug that caused unpredictable behaviour when printing string values with
  the `debug.print` function. The issue was caused by non-null-terminated strings.
```
